lib/prefix.h

```diff
@@ -303,8 +303,8 @@ struct prefix_fs {
 struct prefix_sg {
 	uint8_t family;
 	uint16_t prefixlen;
-	struct in_addr src __attribute__((aligned(8)));
-	struct in_addr grp;
+	struct ipaddr src;
+	struct ipaddr grp;
```

Comment:
I don't think this is a good way of doing this.

We never need mixed IPv4 and IPv6 addresses. We're either compiling `pimd` or `pim6d`, and if we're in `pimd` it's always `in_addr`, and in `pim6d` it's always `in6_addr`. Using `ipaddr` instead prevents the compiler from helping us with warnings when things are confused.

This is especially important for the PIMv6 work because it will be very easy to miss places where the code wasn't updated properly for IPv6.

Reply:
We are going to use the common code for both PIMV4 and PIMV6, that's the reason we have modified in_addr to ipaddr. So any function is using this we are checking for family and will take action accordingly. 

Example: BGP also uses ipaddr for the structure which is used by both IPV4 and IPV6.

In PIM, only difference is we will different daemons for IPV4 and IPV6 using the same data structure.

Reply:
Thanks @eqvinox for your comment. We are using union so that based on v4 or v6 the fields can be used in common code. Since we are going to use in_addr and in6_addr inside the ipaddr union, the compiler will still throw warnings if any.

Reply:
@patrasar @mobash-rasool you misunderstood my comment.

The difference between BGP and PIM is that BGP is one executable that handles everything. For PIM, there are two executables, compiled with two sets of flags and files. `pimd` never needs to handle IPv6. `pim6d` never needs to handle IPv4.

> So any function is using this we are checking for family and will take action accordingly.

That's exactly what I want to avoid. There's a choice here between:

- allowing the data to represent things that are always invalid — i.e. an IPv6 address in `pimd` — relying on checks when the data is accessed, and also adding the possibility of accessing the wrong union member
- removing the ability to express invalid data to begin with, and get extra use out of the compiler's type enforcement

The second option is pretty much always preferable.

Reply:
Yeah David, there had been confusion. We do agree that doing it at compilation time is better. We will do that for other data structures. But prefix_sg is being used between zebra and pim. So zebra will be accessing both IPv4 and IPv6. We will update the PR shortly to express what we are trying to say.

Reply:
NAK, `prefix_sg` is only used in zebra in the VXLAN code and IPMR stats; instead of adding complexity, zebra should use a separate datastructure.

(And `prefix_sg` should be completely moved into pimd and removed from `lib/`.)